client/src/main/java/io/pravega/client/stream/impl/EventStreamReaderImpl.java

```diff
@@ -196,8 +204,7 @@ private PositionInternal getPosition() {
      * have been persisted.
      */
     @GuardedBy("readers")
-    private String updateGroupStateIfNeeded() throws ReaderNotInReaderGroupException {
-        PositionInternal position = (lastPosition == null) ? getPosition() : lastPosition;
+    private String updateGroupStateIfNeeded(PositionInternal position) throws ReaderNotInReaderGroupException {
```

Comment:
Rather than passing this in, we can probably avoid creating it all together. `position` is used in four places:
1. on line 209 where the reader is checkpointing. (This is rare and we can call getPosition() inside of this if without concern)
2. on line 218. This is the same situation as the above. It only occurs on checkpoint.
3. when calling `acquireSegmentsIfNeeded`. However this method doesn't actually use the `position` passed in unless a timer has passed. So perhaps having this method call `shouldAquireSegment` first would avoid the need for position most of the time.
4. `updateLagIfNeeded` here the situation is the same as the above, it's based on a timer. So we could easily have a block that does:
```
if (shouldAquireSegment() || shouldUpdateLag()) {
    PositionImpl position = getPosition();
    // The code currently in lines 231-235 here.
}
```

Reply:
Thanks, I have applied this change (with some additions to that condition).